### DISCUSSION TITLE
The deprecated `clone` function works with Phoenix LiveView, while the new `cloneNode` does not.

### DISCUSSION THREAD

**Discussion by DaTrader** [alpinejs/alpine#4010] [7. Bugs (archived)]
I have tried to replace `Alpine.clone` function with the new `Alpine.cloneNode` in the LiveView-Alpine integration code and virtually nothing seems to works anymore.

```
let liveSocket = new LiveSocket( "/live", Socket, {
  params: { _csrf_token: csrfToken},
  hooks: hooks,
  dom: {
    onBeforeElUpdated( from, to) {
      if ( from._x_dataStack) {
        // window.Alpine.clone( from, to);
        window.Alpine.cloneNode( from, to);
      }
    }
  }
});
```

Can you please un-deprecate the `close` function for the direction in which this is going does not bode well for the projects invested in the LiveView-Alpine integration.

Or am I missing something?

**Comment by Matsa59** [comment]
We had the same issue, currently we're running on the fork for this PR https://github.com/alpinejs/alpine/pull/4000 and our app.js looks like

```js
const liveSocket = new LiveSocket('/live', Socket, {
  dom: {
    onBeforeElUpdated(from, to) {
      if(from && from._x_isInit){
        window.Alpine.cloneNode(from, to);
      }
    }
  },
  // ....
});
```

> [!IMPORTANT]
> `_x_isInit` is introduce in the PR so don't use it if you're not using the fork. 
> Plus it's not guarante this PR to be merged

**Comment by ekwoka** [comment]
Can you specify more about "what doesn't work?"

Or examples?

could using `Alpine.morph` instead be an option?

cloneNode doesn't really do much different.

**Comment by DaTrader** [comment]
It's not like the JS client is crashing, but the behavior of otherwise functioning components is completely different (the behavior contingent on the state in the Alpine variables), so I guess its got to do with resetting the variables or something similar.

I can dig a bit deeper tomorrow, but in the meantime, please check this discussion https://github.com/alpinejs/alpine/pull/4000 that and the reply above by @Matsa59. I haven't yet had the time to read it through as it was just provided but it's the person who suggested switching to cloneNode a couple of days ago in the first place.

**Reply by Matsa59** [reply]
Hum in fact you were right, I got an issue on `x-for`, dunno why the `cloneNode` didn't duplicate the state.

So let me apologies first, and then lets find where the bug comes from ;p

**Reply by Matsa59** [reply]
@ekwoka https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/clone.js#L34 I guess the issue come from the comment above

>  We're hijacking the "walker" so that we only initialize the element we're cloning...

Does this prevent `x-for` to be executed ?

---

I test with this code from the doc 

```html
<ul x-data="{ colors: [
    { id: 1, label: 'Red' },
    { id: 2, label: 'Orange' },
    { id: 3, label: 'Yellow' },
]}">
    <template x-for="color in colors" :key="color.id">
        <li x-text="color.label"></li>
    </template>
</ul>
```

---

## Edit

So I just remove `dontRegisterReactiveSideEffects(() => {` and everything works perfectly. 

Does directives works as reactive ?

**Reply by Matsa59** [reply]
Ok... just spent my afternoon trying to understand what happened and the conclusion is:

For template with x-for (maybe other directives too) Phoenix LV override what alpine did. So if you use any x-for wrap that inside a node that have 'phx-update="ignore"'

I can't believe and didn't though about that...

So from my side I don't see any issue on AlpineJS

---

However it may be interesting that alpine handle changes that could impact "template" and prevent them but it's kind of out of scope of alpine

**Reply by ekwoka** [reply]
I guess, how does phoenix liveview update the page?

Does it morph the Dom or replace elements?

Or is it kind of both?

Both clone and cloneNode seem to try to focus on just copying state over and initializing a specific way.

Since this needs liveview it will probably be much lower on MY list of things to look into (just more difficult to troubleshoot a tool I don't understand).

But knowing at least how liveview does the updates and what this hook is called can help.

Side note though: Generally deprecated means it will only be removed in the next major version.

You are *probably* safe to continue using clone for a while, and at least it will be a clear error of/when Alpine is updated so you can evaluate options then.

**Reply by DaTrader** [reply]
@ekwoka I have a suggestion what we could do if you agree with it. 

I believe we should wait until LiveView 1.0.0 is finally released (should be very soon) so we can be sure about no more breaking changes there. Then I can create a LiveView-Alpine demo app where the issues with `cloneNode` vs `clone` are manifested and you can then try to remedy them. 

Does this sound good to you?

**Reply by ekwoka** [reply]
No guarantees, but it sounds reasonable.
